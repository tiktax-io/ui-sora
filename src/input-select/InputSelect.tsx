import React, { ChangeEvent, forwardRef, HTMLProps, LegacyRef, memo, ReactElement } from 'react'

import { useStyles } from './InputSelect.styles'
import { InputSelectProps, SelectAttributes } from './InputSelect.types'

/******************************************************************************
 * Functional component that takes array of options and render an HTML Select
 * element.
 * @param {CSSProperties} css CSS properties to add to the div wrapping the
 * element.
 * @param {string} [dataTestId] id of the container element to be
 * used on tests as a way to query the component.
 * @param {string} defaultValue the default option selected on render.
 * @property {(value: string) => unknown | void} onChange action to
 * trigger when the selected option has been updated.
 * @param {Array<SelectOption>} options options of input of type select.
 * @returns {ReactElement} Input of type select.
 *****************************************************************************/
const InputSelect = memo(forwardRef(({
  css,
  dataTestId,
  onChange,
  options,
  defaultValue = options[0]
}: InputSelectProps, ref?: LegacyRef<HTMLDivElement>): ReactElement => {
  const classes = useStyles()

  /****************************************************************************
   * HTML attributes and props to be passed to Top parent div element.
   ***************************************************************************/
  const containerProps: HTMLProps<HTMLDivElement> = {
    className: classes.container,
    ref: ref,
    style: css
  }

  /******************************************************************************
   * Attributes that can be passed into HTML Select element.
   *****************************************************************************/
  const selectProps: SelectAttributes = {
    className: classes.select,
    'data-testid': dataTestId,
    id: 'coins',
    onChange: (event: ChangeEvent<HTMLSelectElement>) => {
      event.preventDefault()
      onChange((event.target as HTMLSelectElement).value)
    },
    value: defaultValue
  }

  /******************************************************************************
   * Infer attributes to pass to specific option element.
   * @param {string} option string containing the value of the option.
   * @returns {HTMLProps<HTMLOptionElement>} option element attributes.
   *****************************************************************************/
  const getOptionAttributes = (option: string):
    HTMLProps<HTMLOptionElement> => {
    const attributes: HTMLProps<HTMLOptionElement> = {
      className: classes.option,
      value: option
    }
    return attributes
  }

  return (
    <div {...containerProps}>
      <select {...selectProps}>
        {options.map((option: string, i) => (
          <option key={i} {...getOptionAttributes(option)}>
            {option}
          </option>
        ))}
      </select>
    </div>
  )
}))

export default InputSelect
