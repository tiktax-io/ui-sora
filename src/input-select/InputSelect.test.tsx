import React from 'react'
import { render, screen } from '@testing-library/react'
import InputSelect from './InputSelect'

describe('Block of tests', () => {
  test('renders children text', () => {
    render(<InputSelect>Hola</InputSelect>)
    const demoElement = screen.getByText(/Hola/i)
    expect(demoElement).toBeInTheDocument()
  })
})
