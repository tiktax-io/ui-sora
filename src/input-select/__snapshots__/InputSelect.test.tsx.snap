// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputSelect input select element Can be customised via props named "css". 1`] = `
<div
  className="container-0-2-1"
  style={
    Object {
      "background": "red",
    }
  }
>
  <select
    className="select-0-2-2"
    id="coins"
    onChange={[Function]}
    value="one"
  >
    <option
      className="option-0-2-3"
      value="one"
    >
      one
    </option>
    <option
      className="option-0-2-3"
      value="two"
    >
      two
    </option>
    <option
      className="option-0-2-3"
      value="three"
    >
      three
    </option>
  </select>
</div>
`;

exports[`InputSelect input select element Default component matches snapshot. 1`] = `
<div
  className="container-0-2-1"
>
  <select
    className="select-0-2-2"
    id="coins"
    onChange={[Function]}
    value="test"
  >
    <option
      className="option-0-2-3"
      value="test"
    >
      test
    </option>
    <option
      className="option-0-2-3"
      value="snapshot"
    >
      snapshot
    </option>
  </select>
</div>
`;
